lib/Sema/CSApply.cpp

```diff
@@ -1233,11 +1233,22 @@ namespace {
                          ConstraintLocatorBuilder locator,
                          bool isImplicit, AccessSemantics semantics) {
       // Determine the declaration selected for this subscript operation.
-      auto selected = getOverloadChoice(
+      auto selected = getOverloadChoiceIfAvailable(
                         cs.getConstraintLocator(
                           locator.withPathElement(
                             ConstraintLocator::SubscriptMember)));
-      auto choice = selected.choice;
+
+      // Handles situation where there was a solution available but it didn't
+      // have a proper overload selected from subscript call, might be because
+      // solver was allowed to return free or unresolved types, which can
+      // happen while running diagnostics on one of the expressions.
+      if (!selected.hasValue()) {
+        cs.TC.diagnose(base->getLoc(), diag::type_of_expression_is_ambiguous)
+             .highlight(base->getSourceRange());
+        return nullptr;
```

Comment:
I think this example illustrates why we probably shouldn't get this far at all, too. I've looked at this bug in the past though and couldn't find a clean fix that avoided getting this far.